syntax = "proto3";
option optimize_for = SPEED;

option go_package = "github.com/luyuancpp/dbprotooption;db";
				
import "google/protobuf/descriptor.proto";

//option id  要从50000开始 50000以内pb内部使用

extend google.protobuf.MessageOptions {
	optional uint32 OptionTableId			= 500000 ;	//表ID
	optional string OptionTableName			= 500001 ;	//表名
	optional string OptionPrimaryKey		= 500002 ;	//主键
	optional string OptionForeignKey		= 500003 ;	//外键
	optional string OptionForeignReferences	= 500005 ;	//外键的关联表
	optional string OptionAutoIncrementKey	= 500006 ;	//自增长的Key
	optional string	OptionBackupTableName	= 500007	;	//备份表的名称
	optional bool	OptionLoadOnly			= 500008	;	//只用于数据加载
	optional string	OptionProcedureName		= 500009	;	//存储过程名称
	optional string	OptionProcedureResult	= 500010	;	//存储过程返回结果的PB结构Name
	optional string	OptionIndex				= 500011	;	//建立索引
	optional string OptionUniqueKey			= 500012	;   //
}

// 描述字段的option
extend google.protobuf.FieldOptions {
	optional uint32 OptionRepeatLimit	= 600000 ;	// repeat 字段限制的个数
	optional string OptionDBFieldExtra	= 600001 ;	// 创建数据库要用到得额外信息，如果没有该属性表示非数据库字段
	optional bool   OptionDBReadOnly	= 600002 ;	// 该字段对DB是否只读
	optional string	OptionDBNamePrefix	= 600003 ;	// 数据库名称前缀
	optional string	OptionDBLoadSql		= 600004 ;	// 数据加载SQL语句
	optional bool	OptionDBNeedEscape	= 600005 ;	// 字段是否需要转义存储
	optional bool	OptionDBProcParam	= 600006 ;	// 是否存储过程参数
	optional string OptionOldName		= 600007 ;   //用来自动修改字段名
	optional string OptionTable			= 600008 ;	// 建表名称
	optional string OptionBackUpTabel	= 600009 ;	//备份表
}

// flag 的值, 表示本条数据的操作状态
enum OPERATE_TYPE {
	OPERATE_NONE   = 0;     //sql none
	OPERATE_INSERT = 1;     //sql insert
	OPERATE_UPDATE = 2;     //sql update
	OPERATE_DELETE = 3;     //sql delete
	OPERATE_LOAD   = 4;     //sql load
	OPERATE_DELETE_CACHE = 5;     //cache need delete
	OPERATE_REPLACE = 6;
	OPERATE_CALL_PROC = 7;		// call procedure
	OPERATE_DELETE_BACKUP =8;	//backup it
}

enum BS_RESULT_TYPE{
	BS_RESULT_NONE		= 0;     //
	BS_RESULT_SUCCESS	= 1;
	BS_RESULT_FAILED	= 2;	
};
